File: TLS Inspector/View Controllers/CertificateTableViewController.m
#import "CertificateTableViewController.h"
#import "TableRowSection.h"
#import "TableRowItem.h"
#import "TitleValueTableViewCell.h"
#import "CertificateReminderManager.h"
#import "SANListTableViewController.h"

@interface CertificateTableViewController ()

@property (strong, nonatomic) NSMutableArray<TableRowSection *> * sections;

@end

@implementation CertificateTableViewController

- (void) viewDidLoad {
    [super viewDidLoad];
    self.sections = [NSMutableArray new];

    self.tableView.estimatedRowHeight = 85.0f;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    [uihelper applyStylesToNavigationBar:self.navigationController.navigationBar];

    [[UIMenuController sharedMenuController] setMenuItems: @[
                                                             [[UIMenuItem alloc] initWithTitle:l(@"Verify") action:@selector(verifyValue:)],
                                                             [[UIMenuItem alloc] initWithTitle:l(@"Share") action:@selector(shareValue:)]]];
    [[UIMenuController sharedMenuController] update];

    [self loadCertificate];
    subscribe(@selector(reloadCert), RELOAD_CERT_NOTIFICATION);

    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction target:self action:@selector(showActionSheet:)];
    ADD_SET_THEME_WORKAROUND
}

IMPL_SET_THEME_WORKAROUND

- (void) reloadCert {
    self.sections = [NSMutableArray new];
    [self loadCertificate];
    [self.tableView reloadData];
}

- (void) loadCertificate {
    self.title = selectedCertificate.summary;

    // Subject
    TableRowSection * subjectSection = [TableRowSection sectionWithTitle:@"Subject"];
    NSMutableArray<TableRowItem *> * items = [self rowsFromNameObject:selectedCertificate.subject];
    if (selectedCertificate.extendedValidation) {
        [items addObject:[TableRowItem itemWithTitle:l(@"Extended Validation Authority") value:selectedCertificate.extendedValidationAuthority style:CertificateTableRowItemStyleExpandedValue]];
    }
    subjectSection.items = items;
    [self.sections addObject:subjectSection];

    // Issuer
    TableRowSection * issuerSection = [TableRowSection sectionWithTitle:@"Issuer"];
    issuerSection.items = [self rowsFromNameObject:selectedCertificate.issuer];
    [self.sections addObject:issuerSection];

    // Validity Period
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm"];
    TableRowSection * dateSection = [TableRowSection sectionWithTitle:@"Validity Period"];
    NSMutableArray<TableRowItem *> * dateItems = [NSMutableArray arrayWithArray:@[
                                                                                             [TableRowItem itemWithTitle:l(@"Not Valid Before") value:[dateFormatter stringFromDate:[selectedCertificate notBefore]] style:CertificateTableRowItemStyleBasicValue],
                                                                                             [TableRowItem itemWithTitle:l(@"Not Valid After") value:[dateFormatter stringFromDate:[selectedCertificate notAfter]] style:CertificateTableRowItemStyleBasicValue],
                                                                                             ]];
    dateSection.items = dateItems;
    [self.sections addObject:dateSection];

    // Key Usage
    TableRowSection * keyUsageSection = [TableRowSection sectionWithTitle:@"Key Usage"];
    NSMutableArray<NSString *> * keyUsage = [NSMutableArray arrayWithCapacity:selectedCertificate.keyUsage.count];
    NSMutableArray<TableRowItem *> * usageItems = [NSMutableArray new];
    for (NSString * usage in selectedCertificate.keyUsage) {
        [keyUsage addObject:[lang key:[NSString stringWithFormat:@"keyUsage::%@", usage]]];
    }
    if (keyUsage.count > 0) {
        [usageItems addObject:[TableRowItem itemWithTitle:@"Basic" value:[keyUsage componentsJoinedByString:@", "] style:CertificateTableRowItemStyleExpandedValue]];
    }
    NSMutableArray<NSString *> * extKeyUsage = [NSMutableArray arrayWithCapacity:selectedCertificate.extendedKeyUsage.count];
    for (NSString * usage in selectedCertificate.extendedKeyUsage) {
        [extKeyUsage addObject:[lang key:[NSString stringWithFormat:@"keyUsage::%@", usage]]];
    }
    if (extKeyUsage.count > 0) {
        [usageItems addObject:[TableRowItem itemWithTitle:@"Extended" value:[extKeyUsage componentsJoinedByString:@", "] style:CertificateTableRowItemStyleExpandedValue]];
    }
    keyUsageSection.items = usageItems;
    if (usageItems.count > 0) {
        [self.sections addObject:keyUsageSection];
    }

    // Features
    NSArray<NSString *> * tlsFeatures = selectedCertificate.tlsFeatures;
    if (tlsFeatures != nil && tlsFeatures.count > 0) {
        TableRowSection * tlsFeaturesSection = [TableRowSection sectionWithTitle:@"TLS Features"];
        NSMutableArray<TableRowItem *> * featureItems = [NSMutableArray arrayWithCapacity:tlsFeatures.count];
        for (NSString * feature in tlsFeatures) {
            [featureItems addObject:[TableRowItem itemWithTitle:[lang key:feature] value:@"" style:CertificateTableRowItemStyleBasic]];
        }
        tlsFeaturesSection.items = featureItems;
        [self.sections addObject:tlsFeaturesSection];
    }

    // Public Key Info
    TableRowSection * publicKeySection = [TableRowSection sectionWithTitle:@"Public Key"];
    publicKeySection.items = @[
                               [TableRowItem itemWithTitle:l(@"Algorithm") value:[lang key:[NSString stringWithFormat:@"KeyAlgorithm::%@", selectedCertificate.publicKey.algroithm]] style:CertificateTableRowItemStyleBasicValue],
                               [TableRowItem itemWithTitle:l(@"Signature") value:[lang key:[NSString stringWithFormat:@"CertAlgorithm::%@", selectedCertificate.signatureAlgorithm]] style:CertificateTableRowItemStyleBasicValue],
                               [TableRowItem itemWithTitle:l(@"Size") value:[NSString stringWithFormat:@"%i", selectedCertificate.publicKey.bitLength] style:CertificateTableRowItemStyleBasicValue],
                               ];
    [self.sections addObject:publicKeySection];

    // Fingerprint
    TableRowSection * fingerprintSection = [TableRowSection sectionWithTitle:@"Fingerprints"];
    NSMutableArray<TableRowItem *> * fingerprintSectionItems = [NSMutableArray new];
    if (UserOptions.currentOptions.showFingerprintMD5) {
        [fingerprintSectionItems addObject:[TableRowItem itemWithTitle:@"MD5" value:selectedCertificate.MD5Fingerprint style:CertificateTableRowItemStyleFixedValue]];
    }
    if (UserOptions.currentOptions.showFingerprintSHA128) {
        [fingerprintSectionItems addObject:[TableRowItem itemWithTitle:@"SHA-128" value:selectedCertificate.SHA1Fingerprint style:CertificateTableRowItemStyleFixedValue]];
    }
    if (UserOptions.currentOptions.showFingerprintSHA256) {
        [fingerprintSectionItems addObject:[TableRowItem itemWithTitle:@"SHA-256" value:selectedCertificate.SHA256Fingerprint style:CertificateTableRowItemStyleFixedValue]];
    }
    if (UserOptions.currentOptions.showFingerprintSHA512) {
        [fingerprintSectionItems addObject:[TableRowItem itemWithTitle:@"SHA-512" value:selectedCertificate.SHA512Fingerprint style:CertificateTableRowItemStyleFixedValue]];
    }
    fingerprintSection.items = fingerprintSectionItems;
    if (fingerprintSectionItems.count > 0) {
        [self.sections addObject:fingerprintSection];
    }

    // Subject Alt. Names
    if (selectedCertificate.alternateNames.count > 0) {
        TableRowSection * sanSection = [TableRowSection sectionWithTitle:@"Subject Alternative Names"];
        sanSection.items = @[[TableRowItem itemWithTitle:@"View all alternative names" value:@"" style:CertificateTableRowItemStyleBasicDisclosure]];
        [self.sections addObject:sanSection];
    }
}

- (NSMutableArray<TableRowItem *> *) rowsFromNameObject:(CKNameObject *)name {
    NSMutableArray<TableRowItem *> * subjectItems = [NSMutableArray arrayWithCapacity:8];

    if (name.commonName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::CN") value:name.commonName style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.organizationalUnitName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::OU") value:name.organizationalUnitName style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.organizationName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::O") value:name.organizationName style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.localityName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::L") value:name.localityName style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.stateOrProvinceName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::S") value:name.stateOrProvinceName style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.countryName != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::C") value:l(nstrcat(@"Country::", name.countryName)) style:CertificateTableRowItemStyleBasicValue]];
    }

    if (name.emailAddress != nil) {
        [subjectItems addObject:[TableRowItem itemWithTitle:l(@"Subject::E") value:name.emailAddress style:CertificateTableRowItemStyleBasicValue]];
    }

    return subjectItems;
}

- (void) showActionSheet:(id)sender {
    [uihelper
     presentActionSheetInViewController:self
     attachToTarget:[ActionTipTarget targetWithBarButtonItem:sender]
     title:selectedCertificate.summary
     subtitle:nil
     cancelButtonTitle:[lang key:@"Cancel"]
     items:@[
             l(@"Share Certificate"),
             l(@"Add Certificate Expiry Reminder"),
             ]
     dismissed:^(NSInteger itemIndex) {
         if (itemIndex == 0) {
             [self sharePublicKey:sender];
         } else if (itemIndex == 1) {
             [self addCertificateExpiryReminder:sender];
         }
     }];
}

- (void) sharePublicKey:(UIBarButtonItem *)sender {
    NSData * pem = [selectedCertificate publicKeyAsPEM];
    if (pem) {
        NSString * fileName = format(@"/%@.pem", selectedCertificate.serialNumber);
        NSURL * fileURL = [NSURL fileURLWithPath:[NSTemporaryDirectory() stringByAppendingPathComponent:fileName]];
        [pem writeToURL:fileURL atomically:YES];

        UIActivityViewController *activityController = [[UIActivityViewController alloc]
                                                        initWithActivityItems:@[fileURL]
                                                        applicationActivities:nil];
        activityController.popoverPresentationController.barButtonItem = sender;
        [self presentViewController:activityController animated:YES completion:^() {}];
    } else {
        [uihelper
         presentAlertInViewController:self
         title:l(@"Unable to export certificate")
         body:l(@"We were unable to export the certificate in PEM format.")
         dismissButtonTitle:l(@"Dismiss")
         dismissed:^(NSInteger buttonIndex) {}];
    }
}

- (void) addCertificateExpiryReminder:(UIBarButtonItem *)sender {
    [uihelper
     presentActionSheetInViewController:self
     attachToTarget:[ActionTipTarget targetWithBarButtonItem:sender]
     title:l(@"Notification Date")
     subtitle:l(@"How soon before the certificate expires should we notify you?")
     cancelButtonTitle:l(@"Cancel")
     items:@[
             lv(@"{count} weeks", @[@"2"]),
             l(@"1 month"),
             lv(@"{count} months", @[@"3"]),
             lv(@"{count} months", @[@"6"])]
     dismissed:^(NSInteger itemIndex) {
         if (itemIndex != -1) {
             NSUInteger days = 0;

             switch (itemIndex) {
                 case 0:
                     days = 14;
                     break;
                 case 1:
                     days = 30;
                     break;
                 case 2:
                     days = 60;
                     break;
                 case 3:
                     days = 180;
                     break;
                 default:
                     break;
             }

             [[CertificateReminderManager new]
              addReminderForCertificate:selectedCertificate
              forDomain:currentChain.domain
              daysBeforeExpires:days
              completed:^(NSError *error, BOOL success) {
                  if (success) {
                      [uihelper
                       presentAlertInViewController:self
                       title:l(@"Reminder Added")
                       body:l(@"You can modify the reminder in the reminders app.")
                       dismissButtonTitle:l(@"Dismiss")
                       dismissed:nil];
                  } else if (error) {
                      [uihelper
                       presentErrorInViewController:self
                       error:error
                       dismissed:nil];
                  }
                  // If reminder permission was denied, success = NO and error = Nil
              }];
         }
     }];
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ShowList"]) {
        ((SANListTableViewController *)segue.destinationViewController).items = selectedCertificate.alternateNames;
    }
}

#pragma mark - Table view data source

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return self.sections.count;
}

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)idx {
    TableRowSection * section = self.sections[idx];
    return section.items.count;
}

- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)idx {
    TableRowSection * section = self.sections[idx];
    return l(section.title);
}

- (NSString *) tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)idx {
    TableRowSection * section = self.sections[idx];
    return l(section.footer);
}

- (UITableViewCell *) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TableRowSection * section = self.sections[indexPath.section];
    TableRowItem * row = section.items[indexPath.row];
    return [row cellForRowItem];
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    TableRowSection * section = self.sections[indexPath.section];
    if ([section.title isEqualToString:@"Subject Alternative Names"]) {
        [self performSegueWithIdentifier:@"ShowList" sender:nil];
    }
}

- (BOOL) tableView:(UITableView *)tableView shouldShowMenuForRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (BOOL) tableView:(UITableView *)tableView canPerformAction:(SEL)action forRowAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
    if (action == @selector(copy:)) {
        return YES;
    } else if (action == @selector(shareValue:) || action == @selector(verifyValue:)) {
        TableRowSection * section = self.sections[indexPath.section];
        BOOL match = [section.title isEqualToString:@"Fingerprints"];
        NSLog(@"Section '%@' == 'Fingerprints'", section.title);
        return match;
    }

    return NO;
}

- (void) tableView:(UITableView *)tableView performAction:(SEL)action forRowAtIndexPath:(NSIndexPath *)indexPath withSender:(id)sender {
    if (action == @selector(copy:)) {
        UITableViewCell * cell = [tableView cellForRowAtIndexPath:indexPath];
        NSString * data;
        if ([cell isKindOfClass:[TitleValueTableViewCell class]]) {
            data = ((TitleValueTableViewCell *)cell).valueLabel.text;
        } else {
            data = cell.detailTextLabel.text;
        }
        [[UIPasteboard generalPasteboard] setString:data];
    } else if (action == @selector(shareValue:)) {
        TitleValueTableViewCell * cell = [tableView cellForRowAtIndexPath:indexPath];
        UIActivityViewController *activityController = [[UIActivityViewController alloc]
                                                        initWithActivityItems:@[cell.valueLabel.text]
                                                        applicationActivities:nil];
        activityController.popoverPresentationController.sourceView = cell;
        [self presentViewController:activityController animated:YES completion:nil];
    } else if (action == @selector(verifyValue:)) {
        TitleValueTableViewCell * cell = [tableView cellForRowAtIndexPath:indexPath];
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:l(@"Verify Value")
                                                                                 message:l(@"Enter the value to verify")
                                                                          preferredStyle:UIAlertControllerStyleAlert];
        [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
            textField.placeholder = l(@"Value");
        }];

        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:l(@"Cancel")
                                                               style:UIAlertActionStyleCancel handler:nil];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:l(@"Verify")
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction *action) {
                                                             UITextField * inputField = alertController.textFields.firstObject;
                                                             [self compareValue:inputField.text withValue:cell.valueLabel.text];
                                                         }];

        [alertController addAction:cancelAction];
        [alertController addAction:okAction];

        [self presentViewController:alertController animated:YES completion:nil];
    }
}

- (void) compareValue:(NSString *)left withValue:(NSString *)right {
    NSString * (^formatValue)(NSString *) = ^NSString *(NSString * unformattedValue) {
        return [[unformattedValue lowercaseString] stringByReplacingOccurrencesOfString:@" " withString:@""];
    };
    NSString * formattedCurrentValue = formatValue(left);
    NSString * formattedExpectedValue = formatValue(right);
    if ([formattedExpectedValue isEqualToString:formattedCurrentValue]) {
        [uihelper presentAlertInViewController:self title:l(@"Verified") body:l(@"Both values matched.") dismissButtonTitle:l(@"Dismiss") dismissed:nil];
    } else {
        [uihelper presentAlertInViewController:self title:l(@"Not Verified") body:l(@"Values do not match.") dismissButtonTitle:l(@"Dismiss") dismissed:nil];
    }
}

- (void) shareValue:(UIMenuItem *)sender { }

- (void) verifyValue:(UIMenuItem *)sender { }

@end

File: TLS Inspector/View Controllers/CertificateTableViewController.swift
import UIKit
import CertificateKit

class CertificateTableViewController: UITableViewController {
    var certificate: CKCertificate!
    var sections: [TableViewSection] = []

    override func viewDidLoad() {
        super.viewDidLoad()
        self.buildTable()
    }

    @IBAction func actionButton(_ sender: UIBarButtonItem) {
        UIHelper(self).presentActionSheet(target: ActionTipTarget(barButtonItem: sender),
                                          title: self.certificate.summary,
                                          subtitle: nil,
                                          items: [
                                            lang(key: "Share Certificate"),
                                            lang(key: "Add Certificate Expiry Reminder"),
                                        ])
        { (index) in
            if index == 0 {
                self.shareCertificate(sender)
            } else if index == 1 {
                self.addCertificateReminder(sender)
            }
        }
    }

    func addCertificateReminder(_ sender: UIBarButtonItem) {
        guard let chain = CERTIFICATE_CHAIN else {
            return
        }

        UIHelper(self).presentActionSheet(target: ActionTipTarget(barButtonItem: sender),
                                          title: lang(key: "Notification Date"),
                                          subtitle: lang(key: "How soon before the certificate expires should we notify you?"),
                                          items: [
                                            lang(key: "2 weeks"),
                                            lang(key: "1 month"),
                                            lang(key: "3 months"),
                                            lang(key: "6 months"),
                                        ])
        { (index) in
            var days = 0
            if index == 0 {
                days = 2 * 7
            } else if index == 1 {
                days = 30
            } else if index == 2 {
                days = 30 * 3
            } else if index == 3 {
                days = 30 * 6
            }
            CertificateReminder.addReminder(certificate: self.certificate,
                                            domain: chain.domain,
                                            daysBeforeExpire: days)
            { (rerror) in
                if let error = rerror {
                    UIHelper(self).presentError(error: error, dismissed: nil)
                } else {
                    UIHelper(self).presentAlert(title: lang(key: "Reminder Added"),
                                                body: lang(key: "Use the Reminders app to customize the reminder"),
                                                dismissed: nil)
                }
            }
        }
    }

    func shareCertificate(_ sender: UIBarButtonItem) {
        guard let pem = self.certificate.publicKeyAsPEM else {
            UIHelper(self).presentAlert(title: lang(key: "Unable to export certificate"),
                                        body: lang(key: "We were unable to export the certificate in PEM format."),
                                        dismissed: nil)
            return
        }

        let fileName = (self.certificate.serialNumber ?? "certificate") + ".pem"
        let fileURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName)
        do {
            try pem.write(to: fileURL)
        } catch {
            UIHelper(self).presentError(error: error, dismissed: nil)
            return
        }
        let activityController = UIActivityViewController(activityItems: [fileURL], applicationActivities: nil)
        ActionTipTarget(barButtonItem: sender).attach(to: activityController.popoverPresentationController)
        self.present(activityController, animated: true, completion: nil)
    }

    // swiftlint:disable cyclomatic_complexity
    func buildTable() {
        guard let certificate = CERTIFICATE_CHAIN?.certificates[CURRENT_CERTIFICATE] else {
            return
        }

        self.sections = []
        self.certificate = certificate
        self.title = self.certificate.summary

        if let subjectSection = makeNameSection(name: self.certificate.subject) {
            subjectSection.title = lang(key: "Subject")
            self.sections.append(subjectSection)
        }

        if let issuerSection = makeNameSection(name: self.certificate.issuer) {
            issuerSection.title = lang(key: "Issuer")
            self.sections.append(issuerSection)
        }

        if let validityPeriodSection = makeValidityPeriodSection() {
            self.sections.append(validityPeriodSection)
        }

        if let keyUsageSection = makeKeyUsageSection() {
            self.sections.append(keyUsageSection)
        }

        if let featureSection = makeFeatureSection() {
            self.sections.append(featureSection)
        }

        if let publicKeySection = makePublicKeySection() {
            self.sections.append(publicKeySection)
        }

        if let fingerprintsSection = makeFingerprintsSection() {
            self.sections.append(fingerprintsSection)
        }

        if let keyIdentifierSection = makeKeyIdentifierSection() {
            self.sections.append(keyIdentifierSection)
        }

        if let metadataSection = makeMetadataSection() {
            self.sections.append(metadataSection)
        }

        if let subjectAltNameSection = makeSubjectAltNameSection() {
            self.sections.append(subjectAltNameSection)
        }

        self.tableView.reloadData()
    }
    // swiftlint:enable cyclomatic_complexity

    func makeNameSection(name: CKNameObject) -> TableViewSection? {
        let section = TableViewSection()
        for cn in name.commonNames {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::CN"),
                                                              value: cn,
                                                              useFixedWidthFont: false))
        }
        for country in name.countryCodes {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::C"),
                                                              value: lang(key: "Country::" + country),
                                                              useFixedWidthFont: false))
        }
        for locale in name.states {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::L"),
                                                              value: locale,
                                                              useFixedWidthFont: false))
        }
        for state in name.cities {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::S"),
                                                              value: state,
                                                              useFixedWidthFont: false))
        }
        for org in name.organizations {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::O"),
                                                              value: org,
                                                              useFixedWidthFont: false))
        }
        for ou in name.organizationalUnits {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::OU"),
                                                              value: ou,
                                                              useFixedWidthFont: false))
        }
        for email in name.emailAddresses {
            section.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject::E"),
                                                              value: email,
                                                              useFixedWidthFont: false))
        }
        return section
    }

    func makeValidityPeriodSection() -> TableViewSection? {
        let validitySection = TableViewSection()
        validitySection.title = lang(key: "Validity Period")
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm"

        guard let notBefore = self.certificate.notBefore else {
            return nil
        }
        guard let notAfter = self.certificate.notAfter else {
            return nil
        }

        if let cell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") {
            cell.detailTextLabel?.text = formatter.string(from: notBefore)
            cell.textLabel?.text = lang(key: "Not Before")
            validitySection.cells.append(cell)
        }
        if let cell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") {
            cell.detailTextLabel?.text = formatter.string(from: notAfter)
            cell.textLabel?.text = lang(key: "Not After")
            validitySection.cells.append(cell)
        }

        return validitySection
    }

    func makeKeyUsageSection() -> TableViewSection? {
        let usageSection = TableViewSection()
        usageSection.title = lang(key: "Key Usage")

        var keyUsage: [String] = []
        var extKeyUsage: [String] = []
        for usage in self.certificate.keyUsage ?? [] {
            keyUsage.append(lang(key: "keyUsage::" + usage))
        }
        for usage in self.certificate.extendedKeyUsage ?? [] {
            extKeyUsage.append(lang(key: "keyUsage::" + usage))
        }

        if keyUsage.count == 0 && extKeyUsage.count == 0 {
            return nil
        }

        if keyUsage.count > 0 {
            usageSection.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Basic"),
                                                                   value: keyUsage.joined(separator: ", ")))
        }
        if extKeyUsage.count > 0 {
            usageSection.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Extended"),
                                                                   value: extKeyUsage.joined(separator: ", ")))
        }

        return usageSection
    }

    func makeFeatureSection() -> TableViewSection? {
        let featureSection = TableViewSection()
        featureSection.title = lang(key: "Features")

        guard let features = self.certificate.tlsFeatures else {
            return nil
        }

        for feature in features {
            guard let cell = self.tableView.dequeueReusableCell(withIdentifier: "Basic") else {
                continue
            }
            cell.textLabel?.text = lang(key: feature)
            featureSection.cells.append(cell)
        }

        if featureSection.cells.count == 0 {
            return nil
        }

        return featureSection
    }

    func makeFingerprintsSection() -> TableViewSection? {
        let fingerprintsSection = TableViewSection()
        fingerprintsSection.title = lang(key: "Fingerprints")

        if UserOptions.showFingerprintMD5, let md5 = self.certificate.md5Fingerprint {
            fingerprintsSection.cells.append(TitleValueTableViewCell.Cell(title: "MD5",
                                                                          value: md5,
                                                                          useFixedWidthFont: true))
        }
        if UserOptions.showFingerprintSHA128, let s128 = self.certificate.sha1Fingerprint {
            fingerprintsSection.cells.append(TitleValueTableViewCell.Cell(title: "SHA-128",
                                                                          value: s128,
                                                                          useFixedWidthFont: true))
        }
        if UserOptions.showFingerprintSHA256, let s256 = self.certificate.sha256Fingerprint {
            fingerprintsSection.cells.append(TitleValueTableViewCell.Cell(title: "SHA-258",
                                                                          value: s256,
                                                                          useFixedWidthFont: true))
        }
        if UserOptions.showFingerprintSHA512, let s512 = self.certificate.sha512Fingerprint {
            fingerprintsSection.cells.append(TitleValueTableViewCell.Cell(title: "SHA-512",
                                                                          value: s512,
                                                                          useFixedWidthFont: true))
        }

        if fingerprintsSection.cells.count == 0 {
            return nil
        }
        return fingerprintsSection
    }

    func makePublicKeySection() -> TableViewSection? {
        let pubKeySection = TableViewSection()
        pubKeySection.title = lang(key: "Public Key")

        guard let publicKey = self.certificate.publicKey else {
            return nil
        }

        guard let algorithmCell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") else {
            return nil
        }
        algorithmCell.textLabel?.text = lang(key: "Algoritm")
        algorithmCell.detailTextLabel?.text = lang(key: "KeyAlgorithm::" + publicKey.algroithm)
        pubKeySection.cells.append(algorithmCell)

        guard let signatureCell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") else {
            return nil
        }
        signatureCell.textLabel?.text = lang(key: "Signature")
        let key = "CertAlgorithm::" + (self.certificate.signatureAlgorithm ?? "Unknown")
        signatureCell.detailTextLabel?.text = lang(key: key)
        pubKeySection.cells.append(signatureCell)

        guard let sizeCell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") else {
            return nil
        }
        sizeCell.textLabel?.text = lang(key: "Size")
        sizeCell.detailTextLabel?.text = String.init(format: "%ld", publicKey.bitLength)
        pubKeySection.cells.append(sizeCell)

        return pubKeySection
    }

    func makeKeyIdentifierSection() -> TableViewSection? {
        let keyIdentifierSection = TableViewSection()
        keyIdentifierSection.title = lang(key: "Key Identifier")

        guard let identifiers = self.certificate.keyIdentifiers else {
            return nil
        }

        if let subject = identifiers["subject"] {
            keyIdentifierSection.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Subject"),
                                                                           value: subject,
                                                                           useFixedWidthFont: true))
        }

        if let authority = identifiers["authority"] {
            keyIdentifierSection.cells.append(TitleValueTableViewCell.Cell(title: lang(key: "Authority"),
                                                                           value: authority,
                                                                           useFixedWidthFont: true))
        }

        if keyIdentifierSection.cells.count == 0 {
            return nil
        }
        return keyIdentifierSection
    }

    func makeMetadataSection() -> TableViewSection? {
        let metadataSection = TableViewSection()
        metadataSection.title = lang(key: "Metadata")

        if let serial = self.certificate.serialNumber {
            metadataSection.cells.append(TitleValueTableViewCell.Cell(title: "Serial Number",
                                                                      value: serial,
                                                                      useFixedWidthFont: true))
        }

        if let cell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") {
            cell.textLabel?.text = lang(key: "Certificate Authority")
            cell.detailTextLabel?.text = self.certificate.isCA ? lang(key: "Yes") : lang(key: "No")
            metadataSection.cells.append(cell)
        }

        if let version = self.certificate.version {
            guard let versionCell = self.tableView.dequeueReusableCell(withIdentifier: "Detail") else {
                return nil
            }
            versionCell.textLabel?.text = lang(key: "Version")
            versionCell.detailTextLabel?.text = version.stringValue
            metadataSection.cells.append(versionCell)
        }

        if metadataSection.cells.count == 0 {
            return nil
        }
        return metadataSection
    }

    func makeSubjectAltNameSection() -> TableViewSection? {
        let sanSection = TableViewSection()
        sanSection.title = lang(key: "Subject Alternate Names")
        sanSection.tag = 1

        guard let alternateNames = self.certificate.alternateNames else {
            return nil
        }
        if alternateNames.count == 0 {
            return nil
        }
        guard let cell = self.tableView.dequeueReusableCell(withIdentifier: "Count") else {
            return nil
        }
        guard let label = cell.viewWithTag(1) as? UILabel else {
            return nil
        }
        guard let count = cell.viewWithTag(2) as? UILabel else {
            return nil
        }
        label.text = lang(key: "View All")
        count.text = String.init(format: "%ld", alternateNames.count)
        sanSection.cells.append(cell)

        return sanSection
    }

    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.sections.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.sections[section].cells.count
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return self.sections[section].title
    }

    override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        return self.sections[section].footer
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return self.sections[indexPath.section].cells[indexPath.row]
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let section = self.sections[indexPath.section]
        if section.tag == 1 {
            self.performSegue(withIdentifier: "SANSUNDERTALE", sender: nil)
        }
    }
}
